src/main/java/foundation/privacybydesign/sms/ratelimit/RedisRateLimit.java

```diff
@@ -0,0 +1,245 @@
+            // Ah, there was a request before.
+            if (jedis.exists(key)) {
+                try {
+                    limit = Long.parseLong(jedis.get(key));
+                } catch (NumberFormatException e) {
+                    LOG.error("failed to parse long from " + key + ": " + e.getMessage());
+                }
+            }
+        }
+
+        long startLimit = startLimitIP(now);
+        if (limit < startLimit) {
+            // First visit or previous visit was long ago.
+            // Act like the last try was 3 periods ago.
+            limit = startLimit;
+        }
+
+        // Add a period to the current limit.
+        limit += IP_TIMEOUT;
+        return limit;
+    }
+
+    private long startLimitIP(long now) {
+        return now - IP_TIMEOUT * IP_TRIES;
+    }
+
+    @Override
+    protected synchronized long nextTryPhone(String phone, long now) {
+        // Rate limiter durations (sort-of logarithmic):
+        // 1 10 second
+        // 2 5 minute
+        // 3 1 hour
+        // 4 24 hour
+        // 5+ 1 per day
+        // Keep log 5 days for proper limiting.
+
+        final String key = Redis.createKey(phoneLimitsNamespace, phone);
+
+        Limit limit;
+
+        try (var jedis = pool.getResource()) {
+            limit = limitFromRedis(jedis, key);
+            if (limit == null) {
+                limit = new Limit(now);
+                limitToRedis(jedis, key, limit);
+            }
+        }
+        //
+        // Limit limit = phoneLimits.get(phone);
+        // if (limit == null) {
+        // limit = new Limit(now);
+        // phoneLimits.put(phone, limit);
+        // }
+        long nextTry; // timestamp when the next request is allowed
+        switch (limit.tries) {
+            case 0: // try 1: always succeeds
+                nextTry = limit.timestamp;
+                break;
+            case 1: // try 2: allowed after 10 seconds
+                nextTry = limit.timestamp + 10 * SECOND;
+                break;
+            case 2: // try 3: allowed after 5 minutes
+                nextTry = limit.timestamp + 5 * MINUTE;
+                break;
+            case 3: // try 4: allowed after 3 hours
+                nextTry = limit.timestamp + 3 * HOUR;
+                break;
+            case 4: // try 5: allowed after 24 hours
+                nextTry = limit.timestamp + 24 * HOUR;
+                break;
+            default:
+                throw new IllegalStateException("invalid tries count");
+        }
+        return nextTry;
+    }
+
+    @Override
+    protected synchronized void countIP(String ip, long now) {
+        long nextTry = nextTryIP(ip, now);
+        if (nextTry > now) {
+            throw new IllegalStateException("counting rate limit while over the limit");
+        }
+
+        final String key = Redis.createKey(ipLimitsNamespace, ip);
+        try (var jedis = pool.getResource()) {
+            jedis.set(key, Long.toString(nextTry));
+        }
```

Comment:
In `nextTryIp` a get is done from redis, and here you do a set based on that output. It is wise to do this in a WATCH transaction to prevent that two parallel requests update the value at the same time.

Reply:
It seems like you cannot use the result of a `get` in a `set` further on in the same transaction. The result of the `set` only becomes available after `exec()` is called ([see this](https://github.com/redis/jedis/wiki/AdvancedUsage#transactions))